File: src/Matrix.java
class Matrix {
    private int[] matrix;
    private int width;
    private int height;

    Matrix(int width, int height) {
        matrix = new int[width * height];
        this.height = height;
        this.width = width;
        for (int i = 0; i < width * height; i++) {
            matrix[i] = 0;
        }
    }

    synchronized int get(int y, int x) {
        return matrix[y * width + x];
    }

    synchronized void set(int y, int x, int value) {
        matrix[y * width + x] = value;
    }

    void clear() {
        for (int i = 0; i < width * height; i++) {
            matrix[i] = 0;
        }
    }

    void removeLine(int lineNum) {
        for (int i = 0; i < width; i++) {
            matrix[lineNum * width + i] = 0;
        }
    }

    void moveLine(int src, int dest) {
        System.arraycopy(matrix,src*width,matrix, dest*width,width);
    }
}

File: src/game/Matrix.kt
package game

internal class Matrix(private val width: Int, private val height: Int) {
    private val matrix: IntArray = IntArray(width * height)

    init {
        for (i in 0 until width * height) {
            matrix[i] = 0
        }
    }

    operator fun get(y: Int, x: Int): Int {
        return matrix[y * width + x]
    }

    @Synchronized
    operator fun set(y: Int, x: Int, value: Int) {
        matrix[y * width + x] = value
    }

    fun clear() {
        for (i in 0 until width * height) {
            matrix[i] = 0
        }
    }

    fun removeLine(lineNum: Int) {
        for (i in 0 until width) {
            matrix[lineNum * width + i] = 0
        }
    }

    fun moveLine(src: Int, dest: Int) {
        System.arraycopy(matrix, src * width, matrix, dest * width, width)
    }
}
